src/crypto.ts

```diff
@@ -0,0 +1,155 @@
+const CIPHER_ALGORITHM = 'aes-256-gcm';
+const ENCRYPTION_RESULT_ENCODING = 'base64';
+
+export interface CryptoOptions {
+  encryptionKey: string;
+}
+
+type EncryptInput = string | object | number | boolean;
+
+export interface Crypto {
+  encrypt(input: EncryptInput, aad?: string): Promise<string>;
+  decrypt(output: string | Buffer, aad?: string): Promise<EncryptInput>;
+}
+
+function _validateOpts({ encryptionKey }: CryptoOptions) {
+  if (!encryptionKey) {
+    throw new Error('encryptionKey is required');
+  }
+}
+
+function _validateAAD(aad?: string) {
+  if (aad == null) {
+    return;
+  }
+
+  if (typeof aad !== 'string') {
+    throw new Error('AAD must be a string');
+  }
+
+  if (aad.length === 0) {
+    throw new Error('AAD cannot be an empty string');
+  }
+}
+
+function _generateSalt() {
+  return crypto.randomBytes(SALT_LENGTH_IN_BYTES);
+}
+
+function _generateIV() {
+  return crypto.randomBytes(IV_LENGTH_IN_BYTES);
+}
+
+function _generateKey(encryptionKey: crypto.BinaryLike, salt: string | Buffer): Promise<Buffer> {
+  if (!Buffer.isBuffer(salt)) {
+    salt = Buffer.from(salt, ENCRYPTION_RESULT_ENCODING);
+  }
+
+  return new Promise((resolve, reject) => {
+    crypto.pbkdf2(
+      encryptionKey,
+      salt,
+      KEY_ITERATIONS,
+      KEY_LENGTH_IN_BYTES,
+      KEY_DIGEST,
+      (err, key) => {
+        if (err) {
+          reject(err);
+          return;
+        }
+
+        if (!Buffer.isBuffer(key)) {
+          key = Buffer.from(key, 'binary');
+        }
+
+        resolve(key);
+      }
+    );
+  });
+}
+
+async function _serialize(serializable: any): Promise<string> {
+  const serializedObj = JSON.stringify(serializable);
+  if (serializedObj === undefined) {
+    throw Error('Object to be encrypted must be serializable');
+  }
+  return serializedObj;
+}
+
+/**
+ * Implmenetation of encrypt() and decrypt() taken from https://gist.github.com/AndiDittrich/4629e7db04819244e843,
+ * which was recommended by @jaymode
+ */
+export default function makeCryptoWith(opts: CryptoOptions): Crypto {
+  _validateOpts(opts);
+  const { encryptionKey } = opts;
+  return {
+    async encrypt(input, aad) {
+      _validateAAD(aad);
+      const salt = _generateSalt();
+
+      return Promise.all([
+        _serialize(input),
+        _generateIV(),
+        _generateKey(encryptionKey, salt),
+      ]).then(results => {
+        const [serializedInput, iv, key] = results;
+        const cipher = crypto.createCipheriv(CIPHER_ALGORITHM, key, iv);
+
+        if (aad != null) {
+          cipher.setAAD(Buffer.from(aad, 'utf8'));
+        }
+
+        const encrypted = Buffer.concat([cipher.update(serializedInput, 'utf8'), cipher.final()]);
+        const tag = cipher.getAuthTag();
+
+        return Buffer.concat([salt, iv, tag, encrypted]).toString(ENCRYPTION_RESULT_ENCODING);
+      });
+    },
+    async decrypt(output, aad) {
+      _validateAAD(aad);
+      const outputBytes = Buffer.isBuffer(output)
+        ? output
```

Comment:
In the JS implementation, this was:

```
const outputBytes = new Buffer(output, ENCRYPTION_RESULT_ENCODING);
```

As a result, if `output` was already a buffer, `outputBytes` would be a _copy_ of that buffer.

Over here it looks like if `output` is already a buffer we use the same object as `outputBytes` (i.e. not a copy of it).

Not sure if this is significant or not, but thought I'd point it out since it's a subtle difference from the JS implementation. Might be safer to keep the same implementation and do the following over here instead?

```
const outputBytes = Buffer.from(output, ENCRYPTION_RESULT_ENCODING);
```

If `output` is already a buffer, the second argument should be ignored and a copy of `output` should be stored in `outputBytes`, which would match the JS implementation.

Reply:
`Buffer.from` expects a string so typescript will fail if `output` is either a buffer or a string. Checking for the buffer first should not change the end result from what we had. `new Buffer` has security issues and is deprecated now.